packages/medusa/src/repositories/product.ts

```diff
@@ -314,4 +323,60 @@ export class ProductRepository extends Repository<Product> {
 
     return this.findByIds(productIds)
   }
+
+  public async getFreeTextSearchResultsAndCount(
+    q: string,
+    options: CustomOptions = { where: {} },
+    relations: (keyof Product)[] = []
+  ): Promise<[Product[], number]> {
+    const cleanedOptions = this._cleanOptions(options)
+
+    let qb = this.createQueryBuilder("product")
+      .leftJoinAndSelect("product.variants", "variant")
+      .leftJoinAndSelect("product.collection", "collection")
+      .select(["product.id"])
+      .where(cleanedOptions.where)
+      .andWhere(
+        new Brackets((qb) => {
+          qb.where(`product.description ILIKE :q`, { q: `%${q}%` })
+            .orWhere(`product.title ILIKE :q`, { q: `%${q}%` })
+            .orWhere(`variant.title ILIKE :q`, { q: `%${q}%` })
+            .orWhere(`variant.sku ILIKE :q`, { q: `%${q}%` })
+            .orWhere(`collection.title ILIKE :q`, { q: `%${q}%` })
+        })
+      )
+      .skip(cleanedOptions.skip)
+      .take(cleanedOptions.take)
+
+    if (cleanedOptions.withDeleted) {
+      qb = qb.withDeleted()
+    }
+
+    const [results, count] = await qb.getManyAndCount()
+
+    const products = await this.findWithRelations(
+      relations,
+      results.map((r) => r.id),
+      cleanedOptions.withDeleted
+    )
+
+    return [products, count]
+  }
+
+  private _cleanOptions(
+    options: CustomOptions
+  ): WithRequiredProperty<CustomOptions, "where"> {
+    const where = options.where ?? {}
+    if ("description" in where) {
```

Comment:
suggestion: you can simplify that like this:
```ts
  const where = options.where ?? {}
  delete where.description
  delete where.title
  return { ...options, where }
```

Reply:
I tried that, but unfortunately, typescript is not too happy about that. This is the error I get:
```
Property 'description' does not exist on type '(string | ObjectLiteral | FindConditions<Product> | FindConditions<Product>[] | undefined) & { tags?: FindOperator<ProductTag> | undefined; price_list_id?: FindOperator<...> | undefined; }'.
  Property 'description' does not exist on type 'string & { tags?: FindOperator<ProductTag> | undefined; price_list_id?: FindOperator<PriceList> | undefined; }'.
```

Let me know if we can do anything to make it work nicely with your suggestion 😄

Reply:
don't worry about this suggestion 😂 this is because the consumer always return a query from `buildQuery` which is an object and it is passed to the repositories method, but never a string or undefined like `CustomOptions` suggest. it is in fact `Partial<Writable<TEntity>>` but here the type is more generic and include `string | undefined` and some others. So for now it is not very important to get that micro improvement.